uqmi: support split-APN dual-stack connection

Leo-PL opened the pull request:
On QMI devices, it is possible to establish separate IPv4 and IPv6 PDN sessions using different APN names, without explicit use of QMAP multiplexing. This is due to the fact, that underside normal IPv4v6 APN, two separate PDN sessions are established using same name, but this name can be separate for both.

This is useful to get full dual-stack connection for at least one network, Orange Poland, using APNs "internet" and "internetipv6".
I'm posting this as RFC thanks to @bmork's idea - I'm using this configuration successfully on Sierra EM7455 and serveral ZTE MF286(D) devices - so I think it might be worth upstreaming, or just discussing if there is a better way of implementing this, or even implementing full multiplexing. Currently I don't know of a way to accomplish the same with ModemManager support, even though it does support PDN multiplexing now.

This PR includes #10912 as its base, needing robust IPv6 support in the first place.

Leo-PL committed:
uqmi: support split-APN IPv4 and IPv6 dual-stack

Add two new "v6apn" and "v6profile" properties, to support split-APN
dual-stack onfiguration. This extends the existing ipv4v6 PDP type,
allowing simultaneous connection to two distinct APNs,
one for IPv4 and one for IPv6.
The parameters override existing 'apn' and 'profile' respectively,
if set, but only for IPv6 part of the connection.
If unset, they default to their original values, constituting a standard
IPv4v6 setup.

If a different APN is set for IPv6, a corresponding profile MUST also be
configured, with a different ID, than the IPv4 profile, for example,
profile 2.
Both APNs must match ones configured through QMI or through 'AT+CGDCONT'
command.

Example configuration in UCI:

config interface 'wan'
        option proto 'qmi'
        option device '/dev/cdc-wdm0'
        option autoconnect '1'
        option pdptype 'ipv4v6'
        option apn 'internet'
        option v6apn 'internetipv6'
	option profile '1'
	option v6profile '2'

Corresponding profile configuration:
AT+CGDCONT?
+CGDCONT: 1,"IP","internet","0.0.0.0",0,0,0,0
+CGDCONT: 2,"IPV6","internetipv6","0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0",0,0,0,0

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL committed:
uqmi: do not start 464xlat for dual-stack configurations

If dual-stack configuration is in use, and dhcpv6 option is set, do not start
464xlat sub-interface for dhcpv6 sub-interace , as the configuration already
provides IPv4 connectivty, be it through single or dual APN configuration.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL commented:
ping @bmork

bmork commented:
Looks good to me, although  I have not tested this myself.  I do believe that we should support this special configuration, given that it exists and is supportable.

Leo-PL commented:
If this is merged, I'll prepare support for LuCI as well.

Leo-PL commented:
Ping @dangowrt

dob46 commented:
Tested on MF286D and works without any problems.

Leo-PL commented:
Updated the design and rebased against current tree. @dob46 you might want to test it again, I got rid of the custom `ipv4-and-ipv6` PDP type, you should just switch to `ipv4v6` leaving the rest of configuration the same.

dob46 reviewed (commented):
Works great on Netgear R7800 (MF286D lte modem connected over usbip ).
I suggest to add switch to enable/disable second APN option in Luci

Leo-PL commented:
> Works great on Netgear R7800 (MF286D lte modem connected over usbip ). I suggest to add switch to enable/disable second APN option in Luci

Yeah, I should do that as well - it was my plan to do that after (if) this PR is merged, to have interface set first, but that's unlikely to change, so I might do it in parallel. Maybe in the next week.

Leo-PL commented:
@dob46 I took the plunge and did it earlier, this was quick: https://github.com/openwrt/luci/pull/6426

hauke commented on package/network/utils/uqmi/files/lib/netifd/proto/qmi.sh:
```diff
@@ -309,10 +311,13 @@ proto_qmi_setup() {
 		uqmi -s -d "$device" --set-client-id wds,"$cid_6" --set-ip-family ipv6 > /dev/null 2>&1
 
+		: "${v6apn:=${apn}}"
+		: "${v6profile:=${profile}}"
```
What does the colon do here?

dob46 replied:
I tested with with and without colon and without colon it show 'not found'
`root@OpenWrt:~# LANG=C ./ff.sh

I make script like this
var1=512
var2=256
foa=
: "${foa:=${var1}}"
echo "$foa"
foe=
"${foe:=${var2}}"
echo "$foe"

----------------------------------
with colon
512

without colon
./ff.sh: line 23: 256: not found
256
`

Leo-PL replied:
@hauke it's a no-op, so the default assignment i variable substitution can take place.

hauke replied:
thanks for the explanation, I didn't know this syntax.
